lib/api/confighandler.go

```diff
@@ -0,0 +1,375 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+const configBase = "/rest/config/"
+
+type configHandler struct {
+	*httprouter.Router
+	id  protocol.DeviceID
+	cfg config.Wrapper
+	mut sync.Mutex
+}
+
+func newConfigHandler(id protocol.DeviceID, cfg config.Wrapper) http.Handler {
+	c := &configHandler{
+		Router: httprouter.New(),
+		id:     id,
+		cfg:    cfg,
+		mut:    sync.NewMutex(),
+	}
+
+	c.registerConfig(configBase)
+	c.registerConfigInsync(configBase + "insync")
+	c.registerFolders(configBase + "folders")
+	c.registerDevices(configBase + "devices")
+	c.registerFolder(configBase + "folders/:id")
+	c.registerDevice(configBase + "devices/:id")
+	c.registerOptions(configBase + "options")
+	c.registerLDAP(configBase + "ldap")
+	c.registerGUI(configBase + "gui")
+
+	return c
+}
+
+func (c *configHandler) registerConfig(path string) {
+	legacyPath := "/rest/system/config"
+
+	handler := func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, c.cfg.RawCopy())
+	}
+	c.GET(path, handler)
+	c.GET(legacyPath, handler)
+
+	handler = func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+		c.adjustConfig(w, r)
+	}
+	c.POST(path, handler)
+	c.PUT(path, handler)
+	c.POST(legacyPath, handler)
+	c.PUT(legacyPath, handler)
+}
+
+func (c *configHandler) registerConfigInsync(path string) {
+	legacyPath := "/rest/system/config/insync"
+	handler := func(w http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
+		sendJSON(w, map[string]bool{"configInSync": !c.cfg.RequiresRestart()})
+	}
+	c.GET(path, handler)
+	c.GET(legacyPath, handler)
+}
+
+func (c *configHandler) registerFolders(path string) {
```

Comment:
It would make more sense if the path constants were declared/constructed in these methods, the only reason they are done outside is because of the base path, which can either become a prop of the handler, or be passed in.

This is  to be able to understand whether `/folder/:id` or `/folders` is used for this method.